server/src/main/java/org/opensearch/threadpool/ThreadPool.java

```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
         this(settings, null, customBuilders);
     }
 
+    public void setThreadPool(Settings tpSettings) {
+        Map<String, Settings> tpGroups = tpSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : tpGroups.entrySet()) {
+            String tpName = entry.getKey();
+            Settings tpGroup = entry.getValue();
+            ExecutorHolder holder = executors.get(tpName);
+            assert holder.executor instanceof OpenSearchThreadPoolExecutor;
+            OpenSearchThreadPoolExecutor o = (OpenSearchThreadPoolExecutor) holder.executor;
+            if (holder.info.type == ThreadPoolType.SCALING) {
+                int max = tpGroup.getAsInt("max", o.getMaximumPoolSize());
+                int core = tpGroup.getAsInt("core", o.getCorePoolSize());
+                /*
+                 If we are decreasing, core pool size has to be decreased first.
+                 If we are increasing ,max pool size has to be increased first
+                 This ensures that core pool is always smaller than max pool size .
+                 */
+                if (core < o.getCorePoolSize()) {
+                    o.setCorePoolSize(core);
+                    o.setMaximumPoolSize(max);
+                } else {
+                    o.setMaximumPoolSize(max);
+                    o.setCorePoolSize(core);
+                }
+            } else {
+                int size = tpGroup.getAsInt("size", o.getMaximumPoolSize());
+                if (size < o.getCorePoolSize()) {
+                    o.setCorePoolSize(size);
+                    o.setMaximumPoolSize(size);
+                } else {
+                    o.setMaximumPoolSize(size);
+                    o.setCorePoolSize(size);
+                }
+            }
+        }
+    }
+
+    public void setClusterSettings(ClusterSettings clusterSettings) {
+        this.clusterSettings = clusterSettings;
+        this.clusterSettings.addSettingsUpdateConsumer(CLUSTER_THREAD_POOL_SIZE_SETTING, this::setThreadPool, this::validateSetting);
+    }
+
+    private void validateSetting(Settings tpSettings) {
+        Map<String, Settings> tpGroups = tpSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : tpGroups.entrySet()) {
+            String tpName = entry.getKey();
+            if (THREAD_POOL_TYPES.get(tpName) == null) {
+                throw new IllegalArgumentException("illegal thread_pool name : " + tpName);
+            }
+            Settings tpGroup = entry.getValue();
+            ExecutorHolder holder = executors.get(tpName);
+            assert holder.executor instanceof OpenSearchThreadPoolExecutor;
+            OpenSearchThreadPoolExecutor o = (OpenSearchThreadPoolExecutor) holder.executor;
+            if (holder.info.type == ThreadPoolType.SCALING) {
+                Set<String> so = new HashSet<>(Arrays.asList("max", "core"));
+                if (tpGroup.keySet().stream().allMatch(so::contains) == false) {
```

Comment:
this can be replaced with `so.containsAll(tpGroup.keySet())` for ease of reading. Also, do we need to throw exception? Or can we just log and move on?

Reply:
Done. We need to throw exception as we can't apply that setting .